### PR TITLE
feat: Add support for MongoDB 7

### LINKED ISSUE #8713: Add MongoDB 7 compatbility
### New Feature / Enhancement Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Current Limitation

MongoDB 7 has been released.

### Feature / Enhancement Description

Add Parse Server compatibility for MongoDB 7.

### PR DESCRIPTION
**Pull request by coratgerl**
## Pull Request

This PR add the elements to support the MongoDB v7

Closes: https://github.com/parse-community/parse-server/issues/8713

## Approach
I added the test script in package.json, I ran the script and no one test failed.

## Tasks

- [x] Add tests

### CODE REVIEW

**Review by mtrezza — CHANGES_REQUESTED** [review]
Looks good, thanks for picking this up! Parse Server tests are passing. For full MongoDB 7 support, the MongoDB driver needs to be upgraded as well to the latest version. The current version `"mongodb": "4.10.0"` is only partially compatible with MongoDB 7, see [docs](https://www.mongodb.com/docs/drivers/node/current/compatibility/).

**Comment by coratgerl** [comment]
I updated the version of mongodb to 5.9 to keep the support for node 14.21. I also fix the tests that failed suits to this upgrade. Finally I updated the mongodb-runner to a more recent version :). In local test with mongo v7 works with node 18 and node 14 :)@mtrezza

File: package.json
Comment (mtrezza) [inline package.json]: - Keep the env var fallback `MONGODB_VERSION=${MONGODB_VERSION:=$npm_config_dbversion}`.
- Remove the `mongodb-runner stop --all` from the start command, it's already in `posttest`.

File: package.json
Comment (mtrezza) [inline package.json]: I believe this can be reduced to `mongodb-runner stop --all`?

File: package.json
Comment (mtrezza) [inline package.json]: Is `posttest:mongodb` ever called? If not, we can remove the script.

File: spec/vulnerabilities.spec.js
Comment (mtrezza) [inline spec/vulnerabilities.spec.js:253]: Why has this been removed? It's an important check in this vulnerability test, so before we can remove that we need to understand why.

File: spec/ParseQuery.hint.spec.js
Comment (mtrezza) [inline spec/ParseQuery.hint.spec.js]:
```suggestion
```

File: src/Adapters/Storage/Mongo/MongoCollection.js
Comment (mtrezza) [inline src/Adapters/Storage/Mongo/MongoCollection.js:163]: Does this return a promise? I assume the reason why this was wrapped in a promise previously was that it wasn't returning a promise?

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js
Comment (mtrezza) [inline src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: Why is this needed? Just to make sure we don't introduce a breaking change here.

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js
Comment (mtrezza) [inline src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: Why ` .catch(error => this.handleError(error))` if there is already a `handleError` after this?

**Review by mtrezza — CHANGES_REQUESTED** [review]

**Comment by mtrezza** [comment]
Nice one, thank's for also updating the mongodb-runner, that is overdue.

File: spec/vulnerabilities.spec.js
Comment (coratgerl) [reply spec/vulnerabilities.spec.js:253]: Yes of course, I removed this because in the migration guide of the bson v5 they say this (see screen bellow). Now it is always return Code instances
image

Comment (coratgerl) [reply spec/vulnerabilities.spec.js:253]: https://github.com/mongodb/js-bson/blob/main/docs/upgrade-to-v5.md

File: src/Adapters/Storage/Mongo/MongoCollection.js
Comment (coratgerl) [reply src/Adapters/Storage/Mongo/MongoCollection.js:163]: Yes in the mongo v4 version, they use callback so we needed Promise, but in the mongo v5 the callback is removed so we don't need anymore new Promise :)

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js
Comment (coratgerl) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: It's needed because in some of tests due to invalid id, .collections() don't return an array :). So collections.filter crash because it isn't an array :)

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js
Comment (coratgerl) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: Yes you right :), i revert this change :)

**Comment by coratgerl** [comment]
Thanks for you feedbacks, I fixed them, and make my feedbacks on others :)

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js
Comment (mtrezza) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: > collections() don't return an array

When is that the case, for which test? Is that a behavior change in the MongoDB driver?

File: package.json
Comment (mtrezza) [inline package.json]: We are getting the topology via `MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} `; I believe this should be passed to the runner instead of a fixed `standalone`?

File: package.json
Comment (mtrezza) [inline package.json]: This seens to be missing the topology to be passed to the runner?

File: package.json
Comment (mtrezza) [inline package.json]: I don't think we need to know the mongodb version here?

```suggestion
    "posttest": "mongodb-runner stop --all",
```

File: spec/vulnerabilities.spec.js
Comment (mtrezza) [reply spec/vulnerabilities.spec.js:253]: The test focuses on "BSON type code data"; if the `_bsontype: 'Code',` is removed, how would the MongoDB driver (or DB) know that this is code?

Comment (coratgerl) [reply spec/vulnerabilities.spec.js:253]: Hum, I have a doubt but they this ` The evalFunctions option, when enabled, would return BSON Code typed values as eval-ed JavaScript functions`. For me the evalFunctions always return a Code type. If I explicetely put bsontype to Code we have an Internal error return by mongo 🤔

File: package.json
Comment (mtrezza) [reply package.json]: This was marked as resolved, but it still is missing the topology and version params to be passed.

File: package.json
Comment (mtrezza) [reply package.json]: Also here; why was this resolved?

File: package.json
Comment (mtrezza) [reply package.json]: Why was this resolved?

File: package.json
Comment (coratgerl) [reply package.json]: Lol you are so fast, I worked in local, it's resolved in my local not yet pushed eheh

File: package.json
Comment (coratgerl) [reply package.json]: Lol you are so fast, I worked in local, it's resolved in my local not yet pushed eheh

File: spec/vulnerabilities.spec.js
Comment (mtrezza) [reply spec/vulnerabilities.spec.js:253]: What does the MongoDB driver changelog say?

File: package.json
Comment (mtrezza) [reply package.json]: Please only resolve after the commit has been pushed to this branch, otherwise it's unclear how it has been resolved.

File: spec/vulnerabilities.spec.js
Comment (coratgerl) [reply spec/vulnerabilities.spec.js:253]: They nothing about this juste they upgrade to bson v5. This information is in the migration guide of bson v5

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js
Comment (coratgerl) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: The test are in apple auth and facebook auth. Like 'should use algorithm from key header to verify id_token'.

The problem is thats test spyOn utils.promisify, but now database.collections seems to use it so it's mock this calls also. I search a better solution to mock differently but I don't found anything. With jasmine we can't mock only for onetime, we can't also export the function of facebook adapter and mock it because it's not the same instance. I have no idea. Do you have a proposition @mtrezza  ?

File: package.json
Comment (mtrezza) [inline package.json]: The purpose of these env vars and fallbacks is to use them in the runner:

```suggestion
    "test:mongodb:runnerstart": "cross-env MONGODB_VERSION=${MONGODB_VERSION:=$npm_config_dbversion} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} mongodb-runner start -t ${MONGODB_TOPOLOGY} --version ${MONGODB_VERSION} -- --port 27017",
```

**Review by mtrezza — CHANGES_REQUESTED** [review]

File: spec/vulnerabilities.spec.js
Comment (mtrezza) [reply spec/vulnerabilities.spec.js:253]: > we have an Internal error return by mongo 

Could you post the error?

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js
Comment (mtrezza) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: `spyOn(util, 'promisify')` looks like bad practice; this general helper function can be used anywhere and break the test, as it does here. An alternative would be to look at what it's supposed to mock and mock on a higher level. What should be mocked, the `authUtils`?

**Comment by mtrezza** [comment]
It seems that since you changed the way the query plan results are tested, the tests are now failing for [MongoDB 4.2](https://github.com/parse-community/parse-server/actions/runs/6289806017/job/17076646002?pr=8761). You'd need to introduce an additional test for MongoDB 4.2. Look at the `it_only_mongodb_version` tests to see how this is done with other tests using range operators.

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js
Comment (coratgerl) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: We need to mock this function : 
```javascript
const getFacebookKeyByKeyId = async (keyId, cacheMaxEntries, cacheMaxAge) => {
  const client = jwksClient({
    jwksUri: `${TOKEN_ISSUER}/.well-known/oauth/openid/jwks/`,
    cache: true,
    cacheMaxEntries,
    cacheMaxAge,
  });

  const asyncGetSigningKeyFunction = util.promisify(client.getSigningKey);

  let key;
  try {
    key = await asyncGetSigningKeyFunction(keyId);
  } catch (error) {
    throw new Parse.Error(
      Parse.Error.OBJECT_NOT_FOUND,
      `Unable to find matching key for Key ID: ${keyId}`
    );
  }
  return key;
};
```

File: spec/vulnerabilities.spec.js
Comment (coratgerl) [reply spec/vulnerabilities.spec.js:253]: The only information we have is an error 500 : "Internal server error"

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js
Comment (mtrezza) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: I think we have 3 options:
- a) mock `getFacebookKeyByKeyId` -> not preferred, as we'd be not testing the code in this function
- b) mock `client.getSigningKey`
- c) mock `asyncGetSigningKeyFunction ` -> requires refactoring the code to provide a way to mock it

I'd go with (b).

File: spec/vulnerabilities.spec.js
Comment (mtrezza) [reply spec/vulnerabilities.spec.js:253]: Could you set a breakpoint inside Parse Server to get more details? It's likely an error thrown by the MongoDB driver that gets abstracted by Parse Server.

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js
Comment (coratgerl) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: How can I mock `client.getSigningKey` with jasmine (I am not familiar with it). Moreover the client is created in the function so how can we mock the same instance ?

File: spec/vulnerabilities.spec.js
Comment (mtrezza) [reply spec/vulnerabilities.spec.js:253]: See this discussion for maybe some insight into this: https://github.com/parse-community/parse-server/pull/8560#discussion_r1204953867

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js
Comment (mtrezza) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: See how `spyOn` and `spyOnProperty` is used in the tests

Comment (coratgerl) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: I looked but in this case the problem is that the instant of client in the function is different that every function that we mock. So the mocked function is never called. I don't know if in Jasmine it's possible to mock an internal variable of a function

Comment (mtrezza) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: Could you tell me what are the tests that are failing in which we need to mock this, then I'll take a look

Comment (coratgerl) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: Almost a big part of the facebook test and apple test in AuthenticationAdapter.spec.js

Comment (mtrezza) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: I mean which line in the test specifically?

Comment (mtrezza) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: @coratgerl I hope you didn't get discouraged due to the scope of this PR; it would be great if we could get this ready for merge.

Comment (coratgerl) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: No of course @mtrezza just so busy last days ahah.

This test for example :) 'should use algorithm from key header to verify id_token'. Then we have a bad value in collections here : https://github.com/parse-community/parse-server/blob/d2fda507d9420ec8fc4b32411b832c53419e2cd4/src/Adapters/Storage/Mongo/MongoStorageAdapter.js#L34

**Comment by kantorcodes** [comment]
Would be fantastic to see this merged :)

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js
Comment (coratgerl) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: Do you have a solution for this problem @mtrezza  ? It is the only things missing for merge I think

Comment (mtrezza) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: I just didn't find time yet to look into it, but I'm sure it can be solved

Comment (mtrezza) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: @abhisheksharma010 do you want to look into this? Since you asked to work on this issue

Comment (coratgerl) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: @mtrezza I was able to free up a little time on it, I found a solution that seems to fix the problem :)

**Comment by coratgerl** [comment]
@mtrezza After 1 hour to tryhard to make 4.2 tests passed (I can't run 4.2 tests on my compute, no mongodb-runner available for this version with my arch) all tests passed. For me it's good !

**Comment by mtrezza** [comment]
Great! Let me review this...

File: spec/AuthenticationAdapters.spec.js
Comment (mtrezza) [inline spec/AuthenticationAdapters.spec.js]: I still believe we should not spy on a general method like `promisify`. These test can easily break when some unrelated code is changed in Parse Server. This could make the bugs difficult to identify and cost a lot of time in the future to investigate.

File: src/Adapters/Storage/Mongo/MongoTransform.js
Comment (mtrezza) [inline src/Adapters/Storage/Mongo/MongoTransform.js]: Could you explain this part? This seems to be hacking the MongoDB driver?

Comment (coratgerl) [reply src/Adapters/Storage/Mongo/MongoTransform.js]: I follow the line used by @dblythy here https://github.com/parse-community/parse-server/pull/8560#discussion_r1204953867

File: spec/AuthenticationAdapters.spec.js
Comment (coratgerl) [reply spec/AuthenticationAdapters.spec.js]: I know, but I don't find a better solution. It's not for lack of looking :(

Comment (mtrezza) [reply spec/AuthenticationAdapters.spec.js]: I know you already invested some time into this. Let me see if I can find another solution, otherwise we merge as is.

File: src/Adapters/Storage/Mongo/MongoTransform.js
Comment (mtrezza) [reply src/Adapters/Storage/Mongo/MongoTransform.js]: Got it. The reason the PR wasn't merged was because it wasn't fully clear if that was a breaking change. As @dblythy noted that some types are not supported and may need to be added in the future. I haven't looked deeper into this. @dblythy could you jump in here and maybe help us to make a decision whether this is safe to merge?

File: spec/AuthenticationAdapters.spec.js
Comment (mtrezza) [reply spec/AuthenticationAdapters.spec.js]: Fixed in https://github.com/parse-community/parse-server/pull/8761/commits/e4502a7f3e06e27566460a66a452b2c9b0eb6481

File: src/Adapters/Storage/Mongo/MongoTransform.js
Comment (mtrezza) [reply src/Adapters/Storage/Mongo/MongoTransform.js]: Not sure what the implication of this is, https://github.com/parse-community/parse-server/pull/8560#discussion_r1222366763 is still unclear to me. But since this is an open topic for months without moving ahead, I believe we should take the risk and merge it into Parse Server 7. Since Parse Server 7 will be a major release, developers are supposed to test their app thoroughly before upgrading anyway. So if this really causes a braking change as I understand from https://github.com/parse-community/parse-server/pull/8560#discussion_r1222366763 is possible, then at least it will be with a major release, albeit not documented in the changelog.

**Comment by mtrezza** [comment]
@coratgerl I've resolved the 2 open conversations; could you resolve the conflict, so we can go ahead and merge?

**Comment by coratgerl** [comment]
Thanks you @mtrezza, the conflicts are resolved, we can go ahead !

File: src/Adapters/Storage/Mongo/MongoTransform.js
Comment (Moumouls) [inline src/Adapters/Storage/Mongo/MongoTransform.js]: Hi @mtrezza , i work with @coratgerl , we will take a look about this seems weird to me aslo.

File: package-lock.json
Comment (Moumouls) [inline package-lock.json:1]: we should may be restore package lock from master. Too many change

**Review by Moumouls — CHANGES_REQUESTED** [review]

File: package-lock.json
Comment (coratgerl) [reply package-lock.json:1]: Hey @Moumouls ✋,
The package lock changes is normal since we have changes on the used libraries no ?

File: src/Adapters/Storage/Mongo/MongoTransform.js
Comment (mtrezza) [reply src/Adapters/Storage/Mongo/MongoTransform.js]: Please feel free; I would be fine with merging it into Parse Server 7, i.e. we can merge it on Nov 1st, see https://github.com/parse-community/parse-server/pull/8761#discussion_r1375255686. But any help on this matter is much appreciated.

File: package-lock.json
Comment (mtrezza) [reply package-lock.json:1]: To make sure we have only the changes needed, you could use the lock file from `alpha` and only run `npm i -E mongodb@5.9.0` and `npm i -ED mongodb-runner@5.4.4`. Run this in a npm environment that uses lock file v2, e.g. the npm that comes with Node 16.

Comment (coratgerl) [reply package-lock.json:1]: It's done thanks !

Comment (mtrezza) [reply package-lock.json:1]: Lock changes LGTM

File: src/Adapters/Storage/Mongo/MongoTransform.js
Comment (mtrezza) [reply src/Adapters/Storage/Mongo/MongoTransform.js]: @Moumouls @coratgerl we'll go ahead and merge this for Parse Server 7, unless you still want to investigate this bit.

Comment (Moumouls) [reply src/Adapters/Storage/Mongo/MongoTransform.js]: We need definitely investigate failures without this code @mtrezza before merging, until we have a clear explanation and in case of this code need to be kept a detailed comment about why we have this.

I'll try to take a look this week with @coratgerl , we will do our best !

Comment (mtrezza) [reply src/Adapters/Storage/Mongo/MongoTransform.js]: I agree, that would he great.

File: spec/ParseQuery.Aggregate.spec.js
Comment (mtrezza) [inline spec/ParseQuery.Aggregate.spec.js]: I'm not sure this tests is correct; the title implies "with bsontype", but there is no explicit bsontype being used, it's just a date string?
What's the purpose of `const now` and `const today`, why not just write `new Date().toISOString()`?

**Comment by mtrezza** [comment]
I'm a bit confused by the changes here; in https://github.com/parse-community/parse-server/pull/8761/commits/34015d5aad517237f3a0c8043685fe66b65d7f8b this has been removed:

image

But when I look at the the changes overview, I only see this:

image

**Review by mtrezza — APPROVED** [review]
This looks good, is this ready for merge? I've marked this as breaking, since BSON type support has been removed, even though it was never explicitly documented. Maybe we should still mention it in the changelog, how about:

> Support for BSON type "code" in `Parse.Query` is removed.

**Comment by kantorcodes** [comment]
is there a tag to use mongo 7 support yet?

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [7.0.0-alpha.1](https://github.com/parse-community/parse-server/releases/tag/7.0.0-alpha.1)

**Review by abhisheksharma010 — APPROVED** [review]
Good

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [7.0.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/7.0.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [7.0.0](https://github.com/parse-community/parse-server/releases/tag/7.0.0)

**Pull request merged**
